lib/internal/fs/watch/linux.js

```diff
@@ -0,0 +1,279 @@
+'use strict';
+
+const {
+  SafePromiseAll,
+  PromisePrototypeThen,
+  SafeMap,
+  SafeSet,
+  StringPrototypeStartsWith,
+  StringPrototypeSlice,
+  StringPrototypeLastIndexOf,
+  SymbolAsyncIterator,
+} = primordials;
+
+const { EventEmitter } = require('events');
+const {
+  codes: {
+    ERR_FEATURE_UNAVAILABLE_ON_PLATFORM,
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const { getValidatedPath } = require('internal/fs/utils');
+const { kFSWatchStart } = require('internal/fs/watchers');
+const { kEmptyObject } = require('internal/util');
+const { validateBoolean, validateAbortSignal } = require('internal/validators');
+const path = require('path');
+const { Readable } = require('stream');
+
+let internalAssert;
+let internalSync;
+let internalPromises;
+
+function lazyLoadAssert() {
+  // This is required due to circular dependency
+  internalAssert ??= require('assert');
+  return internalAssert;
+}
+
+function lazyLoadFsPromises() {
+  internalPromises ??= require('fs/promises');
+  return internalPromises;
+}
+
+function lazyLoadFsSync() {
+  internalSync ??= require('fs');
+  return internalSync;
+}
+
+async function traverse(dir, files = new SafeMap()) {
+  const { opendir } = lazyLoadFsPromises();
+
+  const filenames = await opendir(dir);
+  const subdirectories = new SafeSet();
```

Comment:
Is it a trick because of primordials? Because a `Set` doesn't really make sense in this case.

Reply:
No it was not a trick. Do you suggest using arrays?

Reply:
Yes it was to avoid prototype pollution, we could use an array.

Reply:
Yes. All promises added to the set must be different so it's effectively used as an array.

Reply:
I updated the pull request. Thank you for your review @targos